File: packages/client/examples/attach.ts
import { AtmanClient, FetchTransport } from '@atman/client'

const token = process.env.ATMAN_DAEMON_TOKEN
if (!token) throw new Error('ATMAN_DAEMON_TOKEN is required')

const client = await AtmanClient.connect(
  new FetchTransport({ baseUrl: 'http://127.0.0.1:65099', token }),
  { name: 'typescript-example', version: 'example' },
)
const snapshot = await client.command('session.create', {
  request_id: crypto.randomUUID(),
  project_root: process.cwd(),
  title: 'SDK example',
})
const session = await client.attachSession(snapshot.projection.metadata.id)

const unsubscribe = session.subscribe((current) => {
  console.log(
    `cursor=${current.cursor} messages=${current.projection.transcript?.length ?? 0}`,
  )
})
const stop = new AbortController()
const synchronization = session.synchronize({ signal: stop.signal }).catch((error) => {
  if (!stop.signal.aborted) throw error
})

await session.sendMessage('Summarize the current workspace.')
await new Promise((resolve) => setTimeout(resolve, 1_000))
stop.abort()
await synchronization
unsubscribe()

File: crates/atman-client/examples/attach.rs
use atman_client::{Client, ClientIdentity, UnixTransport};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let socket_path = std::env::args()
        .nth(1)
        .ok_or_else(|| std::io::Error::other("usage: attach <path-to-atman.sock>"))?;
    let client = Client::connect(
        UnixTransport::new(socket_path),
        ClientIdentity::new("rust-example", env!("CARGO_PKG_VERSION")),
    )
    .await?;

    let created = client
        .create_session(
            std::env::current_dir()
                .ok()
                .map(|path| path.to_string_lossy().into_owned()),
            Some("SDK example".into()),
        )
        .await?;
    let session_id = created.session_id().clone();
    let session = client.attach_session(session_id).await?;
    let mut states = session.subscribe();

    let synchronizer = session.clone();
    let synchronization = tokio::spawn(async move { synchronizer.synchronize().await });
    session
        .send_message("Summarize the current workspace.", None, Vec::new())
        .await?;
    states.changed().await?;

    let state = states.borrow_and_update();
    println!(
        "session={} cursor={} messages={}",
        session.session_id(),
        state.cursor().0,
        state.projection().transcript.len()
    );
    synchronization.abort();
    Ok(())
}
